lib/multiplexer/wrappers.go

```diff
@@ -155,3 +156,59 @@ func (l *Listener) Close() error {
 	l.cancel()
 	return nil
 }
+
+// PROXYEnabledListener wraps provided listener and can receive and apply PROXY headers and then pass connection up the chain.
+type PROXYEnabledListener struct {
+	cfg Config
+	mux *Mux
+	net.Listener
+}
+
+// NewPROXYEnabledListener creates news instance of PROXYEnabledListener
+func NewPROXYEnabledListener(cfg Config) (*PROXYEnabledListener, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	mux, err := New(cfg) // Creating Mux to leverage protocol detection with PROXY headers
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	pl := &PROXYEnabledListener{
+		cfg:      cfg,
+		mux:      mux,
+		Listener: cfg.Listener,
+	}
+
+	return pl, nil
+}
+
+// Accept gets connection from the wrapped listener and detects whether we receive PROXY headers on it,
+// after first non PROXY protocol detected it returns connection with PROXY addresses applied to it.
+func (p *PROXYEnabledListener) Accept() (net.Conn, error) {
```

Comment:
This `Accept` function will block waiting for I/O from the remote side, blocking whatever accept loop is using it for up to `p.cfg.ReadDeadline`. To fix this you need a buffer, which you would get for free with the regular `multiplexer.Mux`.

Reply:
Right. I've added relying on internal mux for accepting connections, so it wouldn't get blocked if we receive slow connection. We were discussing with @rosstimothy splitting PROXY handling from the multiplexer package, so I would like to keep the PROXYEnabledListener itself.